File: source/shared-json/src/main/java/com/chronos/mobile/source/sharedjson/DefaultTimetableShareCodec.kt
package com.chronos.mobile.source.sharedjson

import com.chronos.mobile.core.model.AcademicConfig
import com.chronos.mobile.core.model.Course
import com.chronos.mobile.core.model.OnlineScheduleEvent
import com.chronos.mobile.core.model.OnlineSchedulePayload
import com.chronos.mobile.core.model.OnlineScheduleWeekDay
import com.chronos.mobile.core.model.Timetable
import com.chronos.mobile.core.model.TimetableImportMetadata
import com.chronos.mobile.core.model.TimetableImportSource
import com.chronos.mobile.core.model.TimetableViewPrefs
import com.chronos.mobile.core.model.currentWeekMonday
import com.chronos.mobile.domain.AcademicCalendarService
import com.chronos.mobile.domain.TimeProvider
import com.chronos.mobile.domain.TimetableShareCodec
import com.chronos.mobile.domain.result.AppError
import com.chronos.mobile.domain.result.AppResult
import com.chronos.mobile.domain.result.appResultOf
import com.chronos.mobile.domain.result.asFailure
import com.chronos.mobile.domain.result.asSuccess
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.time.temporal.TemporalAdjusters
import javax.inject.Inject
import kotlinx.serialization.json.Json
import kotlin.math.abs

class DefaultTimetableShareCodec @Inject constructor(
    private val academicCalendarService: AcademicCalendarService,
    private val timeProvider: TimeProvider,
) : TimetableShareCodec {
    private val json = Json {
        encodeDefaults = true
        explicitNulls = false
        ignoreUnknownKeys = true
        prettyPrint = false
    }

    override fun decode(json: String): AppResult<OnlineSchedulePayload> = appResultOf(
        errorMapper = { AppError.DataFormat("在线课表 JSON 解析失败") },
    ) {
        this.json.decodeFromString(OnlineSchedulePayload.serializer(), json)
    }

    override fun encode(timetable: Timetable): AppResult<String> = appResultOf(
        errorMapper = { AppError.DataFormat("课表导出失败") },
    ) {
        json.encodeToString(OnlineSchedulePayload.serializer(), timetable.toOnlinePayload())
    }

    override fun toTimetable(payload: OnlineSchedulePayload): AppResult<Timetable> {
        val courses = payload.eventList.mapIndexedNotNull { index, event -> event.toCourseOrNull(index) }
        if (courses.isEmpty()) {
            return AppError.Validation("JSON 中未找到可导入的课程数据").asFailure()
        }
        val now = timeProvider.currentTimeMillis()
        val today = timeProvider.today()
        val maxWeek = (payload.weekList.mapNotNull(String::toIntOrNull).maxOrNull()
            ?: courses.flatMap { it.weeks }.maxOrNull()
            ?: 20).coerceAtLeast(20)
        return Timetable(
            id = "online-import",
            name = payload.yearTerm.ifBlank { "在线课表" },
            courses = courses.sortedWith(compareBy<Course> { it.dayOfWeek }.thenBy { it.startPeriod }.thenBy { it.name }),
            createdAt = now,
            updatedAt = now,
            academicConfig = AcademicConfig(
                termStartDate = resolveImportedTermStartDate(payload, today),
                endWeek = maxWeek,
            ),
            importMetadata = TimetableImportMetadata(
                source = payload.importSource.toTimetableImportSourceOrNull() ?: TimetableImportSource.SHARED_JSON,
            ),
            viewPrefs = TimetableViewPrefs(
                showSaturday = courses.any { it.dayOfWeek == 6 },
                showSunday = courses.any { it.dayOfWeek == 7 },
            ),
        ).asSuccess()
    }

    private fun Timetable.toOnlinePayload(): OnlineSchedulePayload {
        val today = timeProvider.today()
        val academicWeek = academicCalendarService.calculateAcademicWeek(today, academicConfig)
        val weekStart = academicCalendarService.resolveWeekStart(
            academicConfig = academicConfig,
            week = academicWeek,
            referenceDate = today,
        )
        val weekNum = academicWeek.toString()
        val weekList = (academicConfig.startWeek..academicConfig.endWeek).map(Int::toString)
        val importSource = importMetadata.source
        return OnlineSchedulePayload(
            yearTerm = name,
            weekNum = weekNum,
            nowMonth = weekStart.monthValue.toString(),
            importSource = importSource.name,
            termStartDate = exportTermStartDate(importSource, academicConfig, today),
            yearTermList = listOf(name),
            weekList = weekList,
            weekDayList = buildWeekDayList(today, weekStart),
            eventList = courses.map { it.toOnlineEvent(weekNum) },
        )
    }

    private fun exportTermStartDate(
        importSource: TimetableImportSource,
        academicConfig: AcademicConfig,
        referenceDate: LocalDate,
    ): String? = when (importSource) {
        TimetableImportSource.ONLINE_EDU -> null
        TimetableImportSource.FILE_HTML,
        TimetableImportSource.SHARED_JSON,
        TimetableImportSource.UNKNOWN -> academicCalendarService.normalizeTermStartDate(academicConfig.termStartDate, referenceDate).toString()
    }

    private fun buildWeekDayList(
        referenceDate: LocalDate,
        weekStart: LocalDate,
    ): List<OnlineScheduleWeekDay> {
        val formatter = DateTimeFormatter.ofPattern("MM/dd")
        return (0..6).map { offset ->
            val date = weekStart.plusDays(offset.toLong())
            OnlineScheduleWeekDay(
                weekDay = when (offset) {
                    0 -> "一"
                    1 -> "二"
                    2 -> "三"
                    3 -> "四"
                    4 -> "五"
                    5 -> "六"
                    else -> "日"
                },
                weekDate = date.format(formatter),
                today = date == referenceDate,
            )
        }
    }

    private fun resolveImportedTermStartDate(
        payload: OnlineSchedulePayload,
        referenceDate: LocalDate,
    ): String {
        val inferred = when (payload.importSource.toTimetableImportSourceOrNull()) {
            TimetableImportSource.ONLINE_EDU -> inferImportedTermStartDate(payload, referenceDate)
            else -> payload.termStartDate
                ?.trim()
                ?.takeIf(String::isNotBlank)
                ?.let(::parseImportedTermStartDate)
                ?: inferImportedTermStartDate(payload, referenceDate)
        }
        return academicCalendarService
            .normalizeTermStartDate((inferred ?: currentWeekMonday(referenceDate)).toString(), referenceDate)
            .toString()
    }

    private fun parseImportedTermStartDate(value: String): LocalDate? =
        runCatching { LocalDate.parse(value) }.getOrNull()

    private fun inferImportedTermStartDate(
        payload: OnlineSchedulePayload,
        referenceDate: LocalDate,
    ): LocalDate? {
        val currentWeek = payload.weekNum.trim().toIntOrNull()?.takeIf { it > 0 } ?: return null
        val weekDays = payload.weekDayList.mapNotNull { it.toImportWeekDayInfo() }
        val anchor = weekDays.firstOrNull { it.dayOfWeek == 1 } ?: weekDays.firstOrNull() ?: return null
        val anchorYear = inferImportWeekDateYear(
            month = anchor.month,
            day = anchor.day,
            yearTerm = payload.yearTerm,
            referenceDate = referenceDate,
        ) ?: return null
        val anchorDate = runCatching { LocalDate.of(anchorYear, anchor.month, anchor.day) }.getOrNull() ?: return null
        val weekStart = anchorDate.minusDays((anchor.dayOfWeek - 1).toLong())
        return weekStart
            .minusWeeks((currentWeek - 1).toLong())
            .with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY))
    }

    private fun OnlineScheduleWeekDay.toImportWeekDayInfo(): ImportWeekDayInfo? {
        val dayOfWeek = weekDay.toImportDayOfWeekOrNull() ?: return null
        val (month, day) = weekDate.toImportMonthDayOrNull() ?: return null
        return ImportWeekDayInfo(dayOfWeek = dayOfWeek, month = month, day = day)
    }

    private fun inferImportWeekDateYear(
        month: Int,
        day: Int,
        yearTerm: String,
        referenceDate: LocalDate,
    ): Int? {
        parseAcademicYears(yearTerm)?.let { (firstYear, secondYear) ->
            val primaryYear = if (month in 8..12) firstYear else secondYear
            if (isValidDate(primaryYear, month, day)) return primaryYear

            val secondaryYear = if (primaryYear == firstYear) secondYear else firstYear
            if (isValidDate(secondaryYear, month, day)) return secondaryYear
        }

        return listOf(referenceDate.year - 1, referenceDate.year, referenceDate.year + 1)
            .mapNotNull { year -> runCatching { LocalDate.of(year, month, day) }.getOrNull() }
            .minByOrNull { candidate -> abs(ChronoUnit.DAYS.between(referenceDate, candidate)) }
            ?.year
    }

    private fun parseAcademicYears(yearTerm: String): Pair<Int, Int>? {
        val match = ACADEMIC_YEAR_PATTERN.find(yearTerm) ?: return null
        val firstYear = match.groupValues.getOrNull(1)?.toIntOrNull() ?: return null
        val secondYear = match.groupValues.getOrNull(2)?.toIntOrNull() ?: return null
        return firstYear to secondYear
    }

    private fun isValidDate(year: Int, month: Int, day: Int): Boolean = runCatching {
        LocalDate.of(year, month, day)
    }.isSuccess

    private fun String.toImportDayOfWeekOrNull(): Int? = when (trim().lowercase()) {
        "1", "一", "周一", "星期一", "mon", "monday" -> 1
        "2", "二", "周二", "星期二", "tue", "tuesday" -> 2
        "3", "三", "周三", "星期三", "wed", "wednesday" -> 3
        "4", "四", "周四", "星期四", "thu", "thursday" -> 4
        "5", "五", "周五", "星期五", "fri", "friday" -> 5
        "6", "六", "周六", "星期六", "sat", "saturday" -> 6
        "7", "日", "天", "周日", "星期日", "sun", "sunday" -> 7
        else -> null
    }

    private fun String.toImportMonthDayOrNull(): Pair<Int, Int>? {
        val parts = trim().split("/")
        if (parts.size != 2) return null
        val month = parts[0].toIntOrNull() ?: return null
        val day = parts[1].toIntOrNull() ?: return null
        if (month !in 1..12 || day !in 1..31) return null
        return month to day
    }

    private fun OnlineScheduleEvent.toCourseOrNull(index: Int): Course? {
        val normalizedName = eventName.normalizedCourseName()
        val dayOfWeek = weekDay.trim().toImportDayOfWeekOrNull() ?: return null
        val startPeriod = sessionStart.trim().toIntOrNull() ?: return null
        if (normalizedName.isBlank()) return null
        val duration = sessionLast.trim().toIntOrNull()?.coerceAtLeast(1)
        val weeks = weekList.mapNotNull(String::toIntOrNull).distinct().sorted()
        val (background, foreground) = coursePalette(normalizedName)
        val endPeriod = sessionList
            .mapNotNull { it.trim().toIntOrNull() }
            .maxOrNull()
            ?: duration?.let { startPeriod + it - 1 }
            ?: startPeriod
        return Course(
            id = eventID.ifBlank { "online-course-${index + 1}" },
            name = normalizedName,
            teacher = memberName.trim(),
            location = address.trim(),
            dayOfWeek = dayOfWeek,
            startPeriod = startPeriod,
            endPeriod = endPeriod.coerceAtLeast(startPeriod),
            color = background,
            textColor = foreground,
            weeks = weeks,
        )
    }

    private fun Course.toOnlineEvent(currentWeek: String): OnlineScheduleEvent = OnlineScheduleEvent(
        weekNum = currentWeek,
        weekDay = dayOfWeek.toString(),
        weekList = weeks.map(Int::toString),
        weekCover = weeks.toWeekCover(),
        sessionList = (startPeriod..endPeriod).map(Int::toString),
        sessionStart = startPeriod.toString(),
        sessionLast = (endPeriod - startPeriod + 1).toString(),
        eventName = name,
        address = location,
        memberName = teacher,
        duplicateGroupType = "none",
        duplicateGroup = 0,
        eventType = "course",
        eventID = id,
    )

    private fun List<Int>.toWeekCover(): String = when {
        isEmpty() -> ""
        size == 1 -> "${first()}周"
        else -> "${first()}-${last()}周"
    }

    private fun String.toTimetableImportSourceOrNull(): TimetableImportSource? =
        TimetableImportSource.entries.firstOrNull { it.name.equals(trim(), ignoreCase = true) }

    private fun coursePalette(name: String): Pair<String, String> =
        COURSE_PALETTE[name.hashCode().mod(COURSE_PALETTE.size)]

    private fun String.normalizedCourseName(): String =
        removePrefix("【调】")
            .removeSuffix("★")
            .removeSuffix("☆")
            .removeSuffix("〇")
            .removeSuffix("■")
            .removeSuffix("◆")
            .trim()

    private data class ImportWeekDayInfo(
        val dayOfWeek: Int,
        val month: Int,
        val day: Int,
    )

    private companion object {
        val ACADEMIC_YEAR_PATTERN = Regex("(20\\d{2})\\D+(20\\d{2})")
        val COURSE_PALETTE = listOf(
            "#EADDFF" to "#21005D",
            "#FFDBC9" to "#311100",
            "#C4EED0" to "#072711",
            "#F9DEDC" to "#410E0B",
            "#D3E3FD" to "#041E49",
            "#FFD8E4" to "#31111D",
        )
    }
}

File: src/lib/parsers/shared-json/default-timetable-share-codec.ts
import { currentWeekMonday } from '$lib/models/defaults';
import type { Course } from '$lib/models/course';
import type {
	OnlineScheduleCampusContext,
	OnlineScheduleEvent,
	OnlineSchedulePayload,
	OnlineScheduleWeekDay
} from '$lib/models/online-schedule';
import { parseOnlineSchedulePayload } from '$lib/models/online-schedule-schema';
import {
	TimetableImportSource,
	normalizeTimetableName,
	type Timetable,
	type TimetableImportMetadata
} from '$lib/models/timetable';
import type { TimetableShareCodec } from '$lib/domain/interfaces/timetable-share-codec';
import { AcademicCalendarService } from '$lib/domain/services/academic-calendar';
import { SystemTimeProvider, type TimeProvider } from '$lib/domain/services/time-provider';
import {
	addDays,
	addWeeks,
	formatIsoDate,
	parseIsoDate,
	previousOrSameMonday,
	safeParseIsoDate
} from '$lib/domain/date';
import { AppError } from '$lib/domain/result/app-error';
import { failure, success, type AppResult } from '$lib/domain/result/app-result';
import { coursePalette, normalizedCourseName } from '$lib/parsers/course-palette';
import { consolidateCourses, sanitizeEventFields } from '$lib/parsers/import-course-utils';

const ACADEMIC_YEAR_PATTERN = /(20\d{2})\D+(20\d{2})/;

interface ImportWeekDayInfo {
	dayOfWeek: number;
	month: number;
	day: number;
}

export class DefaultTimetableShareCodec implements TimetableShareCodec {
	constructor(
		private readonly academicCalendarService = new AcademicCalendarService(),
		private readonly timeProvider: TimeProvider = new SystemTimeProvider()
	) {}

	decode(json: string): AppResult<OnlineSchedulePayload> {
		try {
			return success(parseOnlineSchedulePayload(json));
		} catch {
			return failure(AppError.dataFormat('分享链接解析失败'));
		}
	}

	toTimetable(
		payload: OnlineSchedulePayload,
		campusContext?: OnlineScheduleCampusContext
	): AppResult<Timetable> {
		const yearTerm = payload.yearTerm.trim();
		const courses = consolidateCourses(
			payload.eventList
				.map((event, index) => this.toCourseOrNull(event, index, yearTerm))
				.filter((course): course is Course => course != null)
		);

		if (courses.length === 0) {
			return failure(AppError.validation('分享链接中未找到可导入的课程数据'));
		}

		const now = this.timeProvider.currentTimeMillis();
		const today = this.timeProvider.today();
		const maxWeek = Math.max(
			20,
			Math.max(
				...payload.weekList
					.map((week) => Number.parseInt(week, 10))
					.filter((week) => !Number.isNaN(week)),
				...courses.flatMap((course) => course.weeks)
			)
		);

		const importSource =
			this.toTimetableImportSource(payload.importSource) ?? TimetableImportSource.SHARED_JSON;
		const campusId = campusContext?.campusId;
		const campusPeriodTimes: TimetableImportMetadata['campusPeriodTimes'] | undefined =
			campusContext?.campusPeriodTimes
				? (Object.fromEntries(
						Object.entries(campusContext.campusPeriodTimes).map(([campus, periods]) => [
							campus,
							periods.map((period) => ({ ...period }))
						])
					) as TimetableImportMetadata['campusPeriodTimes'])
				: undefined;
		const periodTimes =
			campusId && campusPeriodTimes
				? (campusPeriodTimes[campusId] ?? []).map((period) => ({ ...period }))
				: [];

		return success({
			id: 'online-import',
			name: normalizeTimetableName(payload.yearTerm),
			courses: [...courses].sort(
				(left, right) =>
					left.dayOfWeek - right.dayOfWeek ||
					left.startPeriod - right.startPeriod ||
					left.name.localeCompare(right.name)
			),
			createdAt: now,
			updatedAt: now,
			academicConfig: {
				termStartDate: this.resolveImportedTermStartDate(payload, today),
				startWeek: 1,
				endWeek: maxWeek,
				periodTimes
			},
			importMetadata: {
				source: importSource,
				campusId,
				campusPeriodTimes
			},
			viewPrefs: {
				showSaturday: courses.some((course) => course.dayOfWeek === 6),
				showSunday: courses.some((course) => course.dayOfWeek === 7),
				showNonCurrentWeekCourses: false
			}
		});
	}

	private resolveImportedTermStartDate(
		payload: OnlineSchedulePayload,
		referenceDate: string
	): string {
		const source = this.toTimetableImportSource(payload.importSource);
		const inferred =
			source === TimetableImportSource.ONLINE_EDU
				? this.inferImportedTermStartDate(payload, referenceDate)
				: ((payload.termStartDate?.trim()
						? this.parseImportedTermStartDate(payload.termStartDate.trim())
						: null) ?? this.inferImportedTermStartDate(payload, referenceDate));

		return this.academicCalendarService.normalizeTermStartDate(
			inferred ?? currentWeekMonday(referenceDate),
			referenceDate
		);
	}

	private parseImportedTermStartDate(value: string): string | null {
		try {
			return formatIsoDate(parseIsoDate(value));
		} catch {
			return null;
		}
	}

	private inferImportedTermStartDate(
		payload: OnlineSchedulePayload,
		referenceDate: string
	): string | null {
		const currentWeek = Number.parseInt(payload.weekNum.trim(), 10);
		if (!currentWeek || currentWeek <= 0) return null;

		const weekDays = payload.weekDayList
			.map((day) => this.toImportWeekDayInfo(day))
			.filter((day): day is ImportWeekDayInfo => day != null);
		const anchor = weekDays.find((day) => day.dayOfWeek === 1) ?? weekDays[0];
		if (!anchor) return null;

		const anchorYear = this.inferImportWeekDateYear(
			anchor.month,
			anchor.day,
			payload.yearTerm,
			referenceDate
		);
		if (!anchorYear) return null;

		const anchorDate = safeParseIsoDate(
			`${anchorYear}-${String(anchor.month).padStart(2, '0')}-${String(anchor.day).padStart(2, '0')}`,
			parseIsoDate(referenceDate)
		);
		const weekStart = addDays(anchorDate, -(anchor.dayOfWeek - 1));
		const termStart = previousOrSameMonday(addWeeks(weekStart, -(currentWeek - 1)));
		return formatIsoDate(termStart);
	}

	private toImportWeekDayInfo(day: OnlineScheduleWeekDay): ImportWeekDayInfo | null {
		const dayOfWeek = this.toImportDayOfWeek(day.weekDay);
		const monthDay = this.toImportMonthDay(day.weekDate);
		if (!dayOfWeek || !monthDay) return null;
		return { dayOfWeek, month: monthDay.month, day: monthDay.day };
	}

	private inferImportWeekDateYear(
		month: number,
		day: number,
		yearTerm: string,
		referenceDate: string
	): number | null {
		const academicYears = this.parseAcademicYears(yearTerm);
		if (academicYears) {
			const [firstYear, secondYear] = academicYears;
			const primaryYear = month >= 8 && month <= 12 ? firstYear : secondYear;
			if (this.isValidDate(primaryYear, month, day)) return primaryYear;
			const secondaryYear = primaryYear === firstYear ? secondYear : firstYear;
			if (this.isValidDate(secondaryYear, month, day)) return secondaryYear;
		}

		const reference = parseIsoDate(referenceDate);
		const candidates = [
			reference.getUTCFullYear() - 1,
			reference.getUTCFullYear(),
			reference.getUTCFullYear() + 1
		]
			.map((year) =>
				safeParseIsoDate(
					`${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`,
					reference
				)
			)
			.filter((date) => date.getUTCMonth() + 1 === month && date.getUTCDate() === day);

		if (candidates.length === 0) return null;
		return candidates
			.reduce((closest, candidate) =>
				Math.abs(candidate.getTime() - reference.getTime()) <
				Math.abs(closest.getTime() - reference.getTime())
					? candidate
					: closest
			)
			.getUTCFullYear();
	}

	private parseAcademicYears(yearTerm: string): [number, number] | null {
		const match = ACADEMIC_YEAR_PATTERN.exec(yearTerm);
		if (!match) return null;
		const firstYear = Number.parseInt(match[1] ?? '', 10);
		const secondYear = Number.parseInt(match[2] ?? '', 10);
		if (Number.isNaN(firstYear) || Number.isNaN(secondYear)) return null;
		return [firstYear, secondYear];
	}

	private isValidDate(year: number, month: number, day: number): boolean {
		const date = new Date(Date.UTC(year, month - 1, day, 12));
		return (
			date.getUTCFullYear() === year &&
			date.getUTCMonth() === month - 1 &&
			date.getUTCDate() === day
		);
	}

	private toImportDayOfWeek(value: string): number | null {
		switch (value.trim().toLowerCase()) {
			case '1':
			case '一':
			case '周一':
			case '星期一':
			case 'mon':
			case 'monday':
				return 1;
			case '2':
			case '二':
			case '周二':
			case '星期二':
			case 'tue':
			case 'tuesday':
				return 2;
			case '3':
			case '三':
			case '周三':
			case '星期三':
			case 'wed':
			case 'wednesday':
				return 3;
			case '4':
			case '四':
			case '周四':
			case '星期四':
			case 'thu':
			case 'thursday':
				return 4;
			case '5':
			case '五':
			case '周五':
			case '星期五':
			case 'fri':
			case 'friday':
				return 5;
			case '6':
			case '六':
			case '周六':
			case '星期六':
			case 'sat':
			case 'saturday':
				return 6;
			case '7':
			case '日':
			case '天':
			case '周日':
			case '星期日':
			case 'sun':
			case 'sunday':
				return 7;
			default:
				return null;
		}
	}

	private toImportMonthDay(value: string): { month: number; day: number } | null {
		const parts = value.trim().split('/');
		if (parts.length !== 2) return null;
		const month = Number.parseInt(parts[0] ?? '', 10);
		const day = Number.parseInt(parts[1] ?? '', 10);
		if (
			Number.isNaN(month) ||
			Number.isNaN(day) ||
			month < 1 ||
			month > 12 ||
			day < 1 ||
			day > 31
		) {
			return null;
		}
		return { month, day };
	}

	private toCourseOrNull(
		event: OnlineScheduleEvent,
		index: number,
		yearTerm: string
	): Course | null {
		const sanitized = sanitizeEventFields(event, yearTerm);
		const normalizedName = normalizedCourseName(sanitized.eventName);
		const dayOfWeek = this.toImportDayOfWeek(sanitized.weekDay);
		const startPeriod = Number.parseInt(sanitized.sessionStart.trim(), 10);
		if (!dayOfWeek || Number.isNaN(startPeriod) || !normalizedName) return null;

		const duration = Number.parseInt(sanitized.sessionLast.trim(), 10);
		const weeks = [
			...new Set(
				sanitized.weekList
					.map((week) => Number.parseInt(week, 10))
					.filter((week) => !Number.isNaN(week))
			)
		].sort((left, right) => left - right);
		const [background, foreground] = coursePalette(normalizedName);
		const sessionMax = sanitized.sessionList
			.map((session) => Number.parseInt(session.trim(), 10))
			.filter((session) => !Number.isNaN(session))
			.reduce((max, session) => Math.max(max, session), Number.NEGATIVE_INFINITY);
		const endPeriod = Number.isFinite(sessionMax)
			? sessionMax
			: !Number.isNaN(duration) && duration >= 1
				? startPeriod + duration - 1
				: startPeriod;

		return {
			id: sanitized.eventID.trim() || `online-course-${index + 1}`,
			name: normalizedName,
			teacher: sanitized.memberName.trim(),
			location: sanitized.address.trim(),
			dayOfWeek,
			startPeriod,
			endPeriod: Math.max(startPeriod, endPeriod),
			color: background,
			textColor: foreground,
			weeks,
			remark: sanitized.remark.trim()
		};
	}

	private toTimetableImportSource(value: string): TimetableImportSource | null {
		const normalized = value.trim().toUpperCase();
		return (Object.values(TimetableImportSource) as string[]).includes(normalized)
			? (normalized as TimetableImportSource)
			: null;
	}
}
